### PR TITLE
Fix upgrade

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: packages/compiler/test/aot/compiler_spec.ts [hunk 1]
```diff
@@ -219,6 +219,55 @@ describe('compiler (unbundled Angular)', () => {
...
+                  declarations: [MyComp]
+                })
+                export class MyModule {}
+              `
+           }
+         };
+         const host = new MockCompilerHost(['/app/app.ts'], FILES, angularFiles);
+         const aotHost = new MockAotCompilerHost(host);
+         let generatedFiles: GeneratedFile[];
+         const warnSpy = spyOn(console, 'warn');
```
Comment (vicb) [inline packages/compiler/test/aot/compiler_spec.ts]: remove

File: packages/compiler/src/metadata_resolver.ts [hunk 2]
```diff
@@ -141,7 +141,26 @@ export class CompileMetadataResolver {
...
+
+  private initComponentFactory(
+      factory: StaticSymbol|ComponentFactory<any>, inputs: {[key: string]: string},
+      outputs: {[key: string]: string}, ngContentSelectors: string[]) {
+    if (!(factory instanceof StaticSymbol)) {
+      for (let propName in inputs) {
+        const templateName = inputs[propName];
+        factory.inputs.push({propName, templateName});
+      }
+      const outputsArr: {propName: string, templateName: string}[] = [];
```
Comment (gkalpak) [inline packages/compiler/src/metadata_resolver.ts]: Unused.

File: packages/upgrade/src/common/downgrade_component.ts [hunk 3]
```diff
@@ -55,28 +49,17 @@ let downgradeCount = 0;
...
- * that appear in the downgraded component's template.
- * These selectors must be provided in the order that they appear in the template as they are
- * mapped by index to the projected nodes.
  *
  * @experimental
  */
-export function downgradeComponent(info: /* ComponentInfo */ {
+export function downgradeComponent(info: {
   component: Type<any>;
+  /** @deprecated This parameter is no longer used */
```
Comment (vicb) [inline packages/upgrade/src/common/downgrade_component.ts]: v4

**Review by vicb — APPROVED** [review]
one nits

File: packages/compiler/test/aot/compiler_spec.ts (hunk 1 above)
Comment (tbosch) [inline packages/compiler/test/aot/compiler_spec.ts]: Done

File: packages/upgrade/src/common/downgrade_component.ts (hunk 3 above)
Comment (tbosch) [inline packages/upgrade/src/common/downgrade_component.ts]: Done

File: packages/upgrade/src/common/downgrade_component.ts [hunk 4]
```diff
@@ -42,11 +41,6 @@ let downgradeCount = 0;
  * "downgraded". This is because once the AoT compiler has run, all metadata about the
```
Comment (gkalpak) [inline packages/upgrade/src/common/downgrade_component.ts]: This last sentence isn't true any more. The necessary metadata is there.

**Review by gkalpak — APPROVED** [review]

File: packages/upgrade/src/common/downgrade_component.ts (hunk 4 above)
Comment (tbosch) [inline packages/upgrade/src/common/downgrade_component.ts]: Done

File: packages/compiler/src/metadata_resolver.ts (hunk 2 above)
Comment (tbosch) [inline packages/compiler/src/metadata_resolver.ts]: Done

**Review by IgorMinar — APPROVED** [review]
Can you mark the last commit as fix or feat rather than refactor since it has user facing impact?

File: packages/examples/upgrade/static/ts/module.ts [hunk 5]
```diff
@@ -145,12 +145,7 @@ ng1AppModule.factory('heroesService', downgradeInjectable(HeroesService));
 
 // #docregion ng2-heroes-wrapper
 // This is directive will act as the interface to the "downgraded"  Angular component
-ng1AppModule.directive(
-    'ng2Heroes',
-    downgradeComponent(
-        // The inputs and outputs here must match the relevant names of the properties on the
-        // "downgraded" component
-        {component: Ng2HeroesComponent, inputs: ['heroes'], outputs: ['addHero', 'removeHero']}));
+ng1AppModule.directive('ng2Heroes', downgradeComponent({component: Ng2HeroesComponent}));
```
Comment (petebacondarwin) [inline packages/examples/upgrade/static/ts/module.ts]: +100

**Comment by tbosch** [comment]
@IgorMinar Good point, done.

**Comment by petebacondarwin** [comment]
LGTM as long as you are sure that:

a) the componentFactory will be initialized with the correct data by the time ngUpgrade needs to use it
b) we can use `Element.matches` to match selectors. (What are these exotic selectors that the compiler will have weeded out - and how would it all work if these exotic selectors were used in a real downgraded component?)

**Comment by tbosch** [comment]
@petebacondarwin good points.
Re 1) As soon as you have the `ComponentFactory`, the compiler is done and all fields are filled correctly.
Re 2) The compiler is still weeding out the exotic selectors, as it does compile the component and with that checks the `selector`s of the `<ng-content>` elements.

**Pull request merged**